### PR TITLE
agent: Fall back to using the Consul agents token when faced with "ACL not found" during deregistrations

### PR DESCRIPTION
**Pull request by weeezes**
### Description

Currently we're seeing an issue when generating Consul ACL tokens for a service with the [Vault Consul Secret Engine](https://www.vaultproject.io/api-docs/secret/consul), where services/checks become "zombies" in the Consul catalog when the Vault token gets revoked or the Consul token gets removed before the service using those tokens has a chance to deregister itself and its checks. The "zombie" services remain in a "healthy" state, even though the TTL checks aren't updated anymore and thus should fall in to failing state.

In case of an "ACL not found" error remove the token associated with the service and check in the state, so that further attemts fall back to using the agent token.

This makes the agent clear the services token from its local state (making the assumption that the token is actually gone and won't ever come back again), and the next re-try falls back to trying out the agent token.

### Testing & Reproduction steps

Our test setup:

* Consul deployed with Helm
* Vault deployed with Helm
* Consul Secret Backend configured to Vault
* A service that registers itself to the local Consul Agent with [Agent.ServiceRegister](https://pkg.go.dev/github.com/hashicorp/consul/api#Agent.ServiceRegister), using a token generated through Vault using the Consul Secret Backend 

When for any reason the Vault token or the generated Consul token gets revoked/removed before the service has a chance to deregister itself there should be services/checks left behind in the servers Catalog, and the agent will log lines like this:

```
2022-09-01T09:41:51.976Z [ERROR] agent.client: RPC failed to server: method=Catalog.Register server=10.42.0.66:8300 error="rpc error making call: ACL not found"
2022-09-01T09:41:51.976Z [WARN]  agent: Check registration blocked by ACLs: check=service:YSbpY7U3HghRGcEijnBEB3 accessorID=
```

When using `DeregisterCriticalServiceAfter` with a TTL check the service and checks do get removed from the agent properly, which is handy because the service doesn't have a working ACL token anymore at that point so that it could update the TTL. But that state doesn't get propagated to the catalog. 

### Links

There are a few issues that I believe are also related to this:

https://github.com/hashicorp/consul/issues/9577
https://github.com/hashicorp/nomad/issues/13537

The response [here](https://github.com/hashicorp/consul/issues/9577#issuecomment-771203024) gave me the idea that this might be a way to fix the problem.

Also this issue seems relevant: https://github.com/hashicorp/consul/issues/4372

### PR Checklist

I might need a bit of help in figuring out what tests to update, and if there's any relevant documentation that should be updated regarding this 😊 .

* [ ] updated test coverage
* [ ] external facing docs updated
* [x] not a security concern

### CODE REVIEW

**Comment by weeezes** [comment]
Hi! Any tips on how to move this forward :)? Is the approach wrong?

**Comment by weeezes** [comment]
Apparently some potentially flaky tests that are causing failures, am I interpreting that right?

Is there something I can do to help this move forward :)?

**Comment by pglass** [comment]
Hi @weeezes, thanks for this change!

If you could do the following, please:

* Update this PR with main so that the tests are re-run
* Add a [changelog entry](https://github.com/hashicorp/go-changelog#change-file-formatting). For this PR, you can add the file `.changelog/14436.txt` (where 14436 is number of this PR) with contents like the following:

  ``````
  $ cat .changelog/14436.txt
  ```release-note:improvement
  agent: client agents fallback to using the agent token for 
  ```
  ``````

Thanks!

File: agent/local/state.go [hunk 1]
```diff
@@ -1295,7 +1295,10 @@ func (l *State) deleteService(key structs.ServiceID) error {
 		l.logger.Info("Deregistered service", "service", key.ID)
 		return nil
 
-	case acl.IsErrPermissionDenied(err), acl.IsErrNotFound(err):
+	case acl.IsErrNotFound(err):
+		l.services[key].Token = ""
+		fallthrough
+	case acl.IsErrPermissionDenied(err):
```
Comment (pglass) [inline agent/local/state.go:1301]: This works!

What I'd suggest actually is not using the service token for the deregistration. That will simplify the logic and behavior slightly.

Like on [line 1273](https://github.com/hashicorp/consul/pull/14436/files#diff-3f2d3e802e18646676be327dbda91428644785290c9d37d5ca37c2124c4170e7R1273), we could do the following (and same for the deleteCheck function)

```diff
- st := l.aclTokenForServiceSync(key, l.tokens.AgentToken)
+ st := l.tokens.AgentToken
```

What do you think abut this? cc @jkirschner-hashicorp @rboyer

**Comment by weeezes** [comment]
Thanks for the comments! I'm on parental leave until early next month, I'll check in more detail when I'm back :).

**Comment by pglass** [comment]
Hi, thanks for this PR! 

After some internal discussion on the approach and in order to include a fix in time for the next Consul release, I've opened a new PR which takes the approach of always using the agent token: https://github.com/hashicorp/consul/pull/16097. I've included your commit from this PR for credit 🙂, and I'll close this PR for now.

**Comment by weeezes** [comment]
awesome, thank you ❤️ !

**Pull request closed without merging**